File: modules/core/src/main/scala/Decision.scala
/*
 * Copyright 2021 Hossein Naderi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package edomata.core

import cats.Applicative
import cats.Eval
import cats.MonadError
import cats.Traverse
import cats.data.Chain
import cats.data.EitherNec
import cats.data.NonEmptyChain
import cats.data.Validated
import cats.data.ValidatedNec
import cats.kernel.Eq

import scala.annotation.tailrec
import scala.annotation.targetName
import scala.util.NotGiven

import Decision._

/** Represents programs that decide in an event driven context
  *
  * This is basically a simple state machine like the following:
  * ```
  * [*] -> InDecisive
  * InDecisive -- event --> Accepted
  * InDecisive -- join --> InDecisive
  * InDecisive -- rejection --> Rejected (resets and terminates)
  * Accepted -- event --> Accepted (accumulates)
  * Accepted -- rejection --> Rejected (resets and terminates)
  * ```
  *
  * It forms a monad error and also is traversable.
  *
  * @tparam R
  *   rejection type
  * @tparam E
  *   event type
  * @tparam A
  *   program output type
  */
sealed trait Decision[+R, +E, +A] extends Product with Serializable { self =>

  /** creates a new decision that changes the output value of this one */
  def map[B](f: A => B): Decision[R, E, B] =
    self match {
      case InDecisive(t)     => InDecisive(f(t))
      case Accepted(evs, a)  => Accepted(evs, f(a))
      case Rejected(reasons) => Rejected(reasons)
    }

  /** binds another decision to this one, creates a new decision */
  def flatMap[R2 >: R, E2 >: E, B](
      f: A => Decision[R2, E2, B]
  ): Decision[R2, E2, B] =
    self match {
      case Accepted(events, result) =>
        f(result) match {
          case Accepted(events2, result) =>
            Accepted(events ++ events2, result)
          case InDecisive(result) => Accepted(events, result)
          case other              => other
        }
      case InDecisive(result) => f(result)
      case Rejected(reasons)  => Rejected(reasons)
    }

  inline def >>=[R2 >: R, E2 >: E, B](
      f: A => Decision[R2, E2, B]
  ): Decision[R2, E2, B] = flatMap(f)

  inline def >>[R2 >: R, E2 >: E, B](
      f: => Decision[R2, E2, B]
  ): Decision[R2, E2, B] = flatMap(_ => f)

  def as[B](b: B): Decision[R, E, B] = map(_ => b)

  def flatTap[R2 >: R, E2 >: E, B](
      f: A => Decision[R2, E2, B]
  ): Decision[R2, E2, A] = flatMap(a => f(a).as(a))

  /** whether is rejected or not */
  def isRejected: Boolean = self match {
    case Rejected(_) => true
    case _           => false
  }

  /** whether is accepted or not */
  def isAccepted: Boolean = self match {
    case Accepted(_, _) => true
    case _              => false
  }

  /** traverses this decision, run fr if there are errors and runs fa if there
    * is some output
    */
  def visit[B](fr: NonEmptyChain[R] => B, fa: A => B): B = self match {
    case Decision.InDecisive(a)  => fa(a)
    case Decision.Accepted(_, a) => fa(a)
    case Decision.Rejected(r)    => fr(r)
  }

  /** Ignores events and creates a ValidatedNec */
  def toValidated: ValidatedNec[R, A] = self match {
    case Decision.Accepted(_, t) => Validated.Valid(t)
    case Decision.InDecisive(t)  => Validated.Valid(t)
    case Decision.Rejected(r)    => Validated.Invalid(r)
  }

  /** Ignores events and errors and creates an Option that contains program
    * output
    */
  def toOption: Option[A] = visit(_ => None, Some(_))

  /** Ignores events and creates an Either */
  def toEither: EitherNec[R, A] = visit(Left(_), Right(_))

  /** Ignores output value */
  def void: Decision[R, E, Unit] = map(_ => ())

  /** Validates output using a ValidatedNec */
  def validate[R2 >: R, B](f: A => ValidatedNec[R2, B]): Decision[R2, E, B] =
    flatMap(a => Decision.validate(f(a)))

  /** Validates output using an EitherNec */
  @targetName("validateEitherNec")
  def validate[R2 >: R, B](f: A => EitherNec[R2, B]): Decision[R2, E, B] =
    flatMap(a => Decision.fromEitherNec(f(a)))

  /** Validates output using an Either */
  @targetName("validateEither")
  def validate[R2 >: R, B](f: A => Either[R2, B]): Decision[R2, E, B] =
    flatMap(a => Decision.fromEither(f(a)))

  /** Asserts output using a ValidatedNec without changing it */
  def assert[R2 >: R, B](f: A => ValidatedNec[R2, B]): Decision[R2, E, A] =
    flatTap(a => Decision.validate(f(a)))

  /** Asserts output using an EitherNec without changing it */
  @targetName("assertEitherNec")
  def assert[R2 >: R, B](f: A => EitherNec[R2, B]): Decision[R2, E, A] =
    flatTap(a => Decision.fromEitherNec(f(a)))

  /** Asserts output using an Either without changing it */
  @targetName("assertEither")
  def assert[R2 >: R, B](f: A => Either[R2, B]): Decision[R2, E, A] =
    flatTap(a => Decision.fromEither(f(a)))

}

object Decision extends DecisionConstructors, DecisionCatsInstances0 {
  final case class InDecisive[T](result: T)
      extends Decision[Nothing, Nothing, T]
  final case class Accepted[E, T](events: NonEmptyChain[E], result: T)
      extends Decision[Nothing, E, T]
  final case class Rejected[R](reasons: NonEmptyChain[R])
      extends Decision[R, Nothing, Nothing]
}

sealed trait DecisionConstructors {

  /** Constructs a program that outputs a pure value */
  def apply[T](t: T): Decision[Nothing, Nothing, T] = InDecisive(t)

  /** Constructs a program that outputs a pure value */
  def pure[T](t: T): Decision[Nothing, Nothing, T] = InDecisive(t)

  /** Constructs a program that outputs a trivial output */
  def unit: Decision[Nothing, Nothing, Unit] = InDecisive(())

  /** Constructs a program that decides to accept a sequence of events */
  def accept[E](ev: E, evs: E*): Decision[Nothing, E, Unit] =
    acceptReturn(())(ev, evs: _*)

  /** Constructs a program that decides to accept a sequence of events and also
    * returns an output
    */
  def acceptReturn[E, T](t: T)(ev: E, evs: E*): Decision[Nothing, E, T] =
    Accepted(NonEmptyChain.of(ev, evs: _*), t)

  /** Constructs a program that decides to accept a sequence of events or do
    * nothing based on a boolean predicate
    */
  def acceptWhen[E](
      predicate: => Boolean
  )(ev: E, evs: E*): Decision[Nothing, E, Unit] =
    if predicate then accept(ev, evs: _*)
    else unit

  /** Constructs a program that decides to accept an optional single event or do
    * nothing
    */
  def acceptWhen[E](ev: Option[E]): Decision[Nothing, E, Unit] =
    ev.fold(unit)(accept(_))

  /** Constructs a program that decides to accept an optional single event or
    * reject otherwise
    */
  def acceptWhen[R, E](
      ev: Option[E]
  )(orElse: R, other: R*): Decision[R, E, Unit] =
    ev.fold(reject(orElse, other: _*))(accept(_))

  /** Constructs a program that decides to accept an optional single event or
    * reject otherwise
    */
  def acceptWhen[R, E](
      eit: Either[R, E]
  ): Decision[R, E, Unit] = eit.fold(reject(_), accept(_))

  /** Constructs a program that decides to accept an optional single event or
    * reject otherwise
    */
  @targetName("acceptWhenEitherNec")
  def acceptWhen[R, E](
      eit: EitherNec[R, E]
  ): Decision[R, E, Unit] = eit.fold(Rejected(_), accept(_))

  /** Constructs a program that decides to reject with a sequence of reasons */
  def reject[R, E](
      reason: R,
      otherReasons: R*
  ): Decision[R, E, Nothing] =
    Rejected(NonEmptyChain.of(reason, otherReasons: _*))

  /** Constructs a program that decides to reject with a sequence of reasons if
    * a predicate satisfies
    */
  def rejectWhen[R](
      predicate: => Boolean
  )(ev: R, evs: R*): Decision[R, Nothing, Unit] =
    if predicate then reject(ev, evs: _*)
    else unit

  /** Constructs a program that decides to reject with an optional single reason
    * or do nothing otherwise
    */
  def rejectWhen[R](ev: Option[R]): Decision[R, Nothing, Unit] =
    ev.fold(unit)(reject(_))

  /** Constructs a program that uses a validation to decide whether to output a
    * value or reject with error(s)
    *
    * You can also use .toDecision syntax for more convenience
    */
  def validate[R, T](
      validation: ValidatedNec[R, T]
  ): Decision[R, Nothing, T] =
    validation match {
      case Validated.Invalid(e) => Rejected(e)
      case Validated.Valid(a)   => InDecisive(a)
    }

  /** Constructs a program from an optional value, that outputs value if exists
    * or rejects otherwise
    *
    * You can also use .toDecision syntax for more convenience
    */
  def fromOption[R, T](
      opt: Option[T],
      orElse: R,
      other: R*
  ): Decision[R, Nothing, T] = opt.fold(reject(orElse, other: _*))(pure(_))

  /** Constructs a program that either outputs a value or rejects
    */
  def fromEither[R, T](
      eit: Either[R, T]
  ): Decision[R, Nothing, T] = eit.fold(reject(_), pure(_))

  /** Constructs a program that either outputs a value or rejects
    *
    * You can also use .toDecision syntax for more convenience.
    */
  def fromEitherNec[R, T](
      eit: EitherNec[R, T]
  ): Decision[R, Nothing, T] = eit.fold(Rejected(_), pure(_))
}

private type D[R, E] = [T] =>> Decision[R, E, T]

sealed trait DecisionCatsInstances0 extends DecisionCatsInstances1 {
  given [R, E]: MonadError[D[R, E], NonEmptyChain[R]] =
    new MonadError[D[R, E], NonEmptyChain[R]] {
      override def pure[A](x: A): Decision[R, E, A] = Decision.pure(x)

      override def map[A, B](
          fa: Decision[R, E, A]
      )(f: A => B): Decision[R, E, B] =
        fa.map(f)

      override def flatMap[A, B](fa: Decision[R, E, A])(
          f: A => Decision[R, E, B]
      ): Decision[R, E, B] =
        fa.flatMap(f)

      @tailrec
      private def step[A, B](a: A, evs: Chain[E] = Chain.empty)(
          f: A => Decision[R, E, Either[A, B]]
      ): Decision[R, E, B] =
        f(a) match {
          case Decision.Accepted(ev2, Left(a)) =>
            step(a, evs ++ ev2.toChain)(f)
          case Decision.Accepted(ev2, Right(b)) =>
            Decision.Accepted(ev2.prependChain(evs), b)
          case Decision.InDecisive(Left(a)) => step(a, evs)(f)
          case Decision.InDecisive(Right(b)) =>
            NonEmptyChain
              .fromChain(evs)
              .fold(Decision.InDecisive(b))(Decision.Accepted(_, b))
          case Decision.Rejected(rs) => Decision.Rejected(rs)
        }

      override def tailRecM[A, B](a: A)(
          f: A => Decision[R, E, Either[A, B]]
      ): Decision[R, E, B] =
        step(a)(f)

      def handleErrorWith[A](
          fa: Decision[R, E, A]
      )(f: NonEmptyChain[R] => Decision[R, E, A]): Decision[R, E, A] =
        fa match {
          case Decision.Rejected(e) => f(e)
          case other                => other
        }

      def raiseError[A](e: NonEmptyChain[R]): Decision[R, E, A] =
        Decision.Rejected(e)

    }

  given [R, E, T]: Eq[Decision[R, E, T]] = Eq.instance(_ == _)
}

sealed trait DecisionCatsInstances1 {

  given [R, E]: Traverse[D[R, E]] = new Traverse {
    def traverse[G[_]: Applicative, A, B](fa: Decision[R, E, A])(
        f: A => G[B]
    ): G[Decision[R, E, B]] = fa.visit(
      rej => Applicative[G].pure(Decision.Rejected(rej)),
      a => Applicative[G].map(f(a))(b => fa.map(_ => b))
    )

    def foldLeft[A, B](fa: Decision[R, E, A], b: B)(f: (B, A) => B): B =
      fa.visit(_ => b, a => f(b, a))

    def foldRight[A, B](fa: Decision[R, E, A], lb: Eval[B])(
        f: (A, Eval[B]) => Eval[B]
    ): Eval[B] =
      fa.visit(_ => lb, a => f(a, lb))
  }
}

private[edomata] transparent trait DecisionSyntax {
  extension [R, T](self: ValidatedNec[R, T]) {
    def toDecision[E]: Decision[R, E, T] = Decision.validate(self)
  }

  private type NonChain[R] = NotGiven[R <:< NonEmptyChain[Nothing]]

  extension [T](self: Option[T]) {
    def toDecision[R](orElse: R, others: R*): Decision[R, Nothing, T] =
      Decision.fromOption(self, orElse, others: _*)
    def toAccepted: Decision[Nothing, T, Unit] = Decision.acceptWhen(self)
    def toAcceptedOr[R](orElse: R, others: R*): Decision[R, T, Unit] =
      Decision.acceptWhen(self)(orElse, others: _*)
    def toRejected: Decision[T, Nothing, Unit] = Decision.rejectWhen(self)
  }

  extension [R, T](self: EitherNec[R, T]) {
    def toDecision: Decision[R, Nothing, T] =
      Decision.fromEitherNec(self)
    def toAccepted: Decision[R, T, Unit] = Decision.acceptWhen(self)
  }

  extension [R, T](self: Either[R, T])(using NonChain[R]) {
    def toDecision: Decision[R, Nothing, T] =
      Decision.fromEither(self)
    def toAccepted: Decision[R, T, Unit] = Decision.acceptWhen(self)
  }

  extension [T](value: T) {
    def asDecision: Decision[Nothing, Nothing, T] = Decision.pure(value)
    def accept: Decision[Nothing, T, Unit] = Decision.accept(value)
    def reject: Decision[T, Nothing, Nothing] = Decision.reject(value)
  }
}

File: rust/crates/edomata-core/src/decision.rs
//! [`Decision`]: the pure state machine that drives every Edomata program.

use std::ops::ControlFlow;

use crate::NonEmpty;

/// Result of a domain decision.
///
/// A `Decision` is a small state machine:
///
/// ```text
/// [*] -> InDecisive
/// InDecisive -- event --> Accepted
/// InDecisive -- join --> InDecisive
/// InDecisive -- rejection --> Rejected (resets and terminates)
/// Accepted -- event --> Accepted (accumulates)
/// Accepted -- rejection --> Rejected (resets and terminates)
/// ```
///
/// It behaves like a monad with error handling: [`Decision::and_then`]
/// accumulates events in order and short-circuits on the first rejection.
///
/// # Type parameters
///
/// * `R` – rejection type
/// * `E` – event type
/// * `A` – program output type
///
/// ```
/// use edomata_core::{Decision, nonempty};
///
/// let d = Decision::accept(1).and_then(|_| Decision::accept(2));
/// assert_eq!(d, Decision::Accepted { events: nonempty![1, 2], result: () });
///
/// let r: Decision<&str, i32, ()> = d.and_then(|_| Decision::reject("no"));
/// assert!(r.is_rejected());
/// ```
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum Decision<R, E, A> {
    /// No decision was made yet; the program produced an output only.
    InDecisive(A),
    /// The program accepted one or more events.
    Accepted {
        /// Events accepted so far, in order.
        events: NonEmpty<E>,
        /// The program output.
        result: A,
    },
    /// The program rejected the request with one or more reasons.
    Rejected(NonEmpty<R>),
}

impl<R, E, A> Decision<R, E, A> {
    // ---------------------------------------------------------------------
    // Constructors
    // ---------------------------------------------------------------------

    /// Constructs a program that outputs a pure value.
    pub fn pure(a: A) -> Self {
        Decision::InDecisive(a)
    }

    /// Constructs a program that decides to accept a sequence of events and
    /// also returns an output.
    pub fn accept_return(result: A, events: impl Into<NonEmpty<E>>) -> Self {
        Decision::Accepted {
            events: events.into(),
            result,
        }
    }

    /// Constructs a program that decides to reject with a sequence of reasons.
    pub fn reject(reasons: impl Into<NonEmpty<R>>) -> Self {
        Decision::Rejected(reasons.into())
    }

    /// Constructs a program that either outputs a value or rejects with all
    /// the given reasons.
    pub fn validate(validation: Result<A, NonEmpty<R>>) -> Self {
        match validation {
            Ok(a) => Decision::InDecisive(a),
            Err(reasons) => Decision::Rejected(reasons),
        }
    }

    /// Alias of [`Decision::validate`].
    pub fn from_result_nec(validation: Result<A, NonEmpty<R>>) -> Self {
        Self::validate(validation)
    }

    /// Constructs a program that either outputs a value or rejects with a
    /// single reason.
    pub fn from_result(result: Result<A, R>) -> Self {
        match result {
            Ok(a) => Decision::InDecisive(a),
            Err(reason) => Decision::Rejected(NonEmpty::new(reason)),
        }
    }

    /// Constructs a program from an optional value, that outputs the value if
    /// it exists or rejects otherwise.
    pub fn from_option(opt: Option<A>, or_else: impl Into<NonEmpty<R>>) -> Self {
        match opt {
            Some(a) => Decision::InDecisive(a),
            None => Decision::Rejected(or_else.into()),
        }
    }

    // ---------------------------------------------------------------------
    // Combinators
    // ---------------------------------------------------------------------

    /// Creates a new decision that changes the output value of this one.
    pub fn map<B, F: FnOnce(A) -> B>(self, f: F) -> Decision<R, E, B> {
        match self {
            Decision::InDecisive(a) => Decision::InDecisive(f(a)),
            Decision::Accepted { events, result } => Decision::Accepted {
                events,
                result: f(result),
            },
            Decision::Rejected(reasons) => Decision::Rejected(reasons),
        }
    }

    /// Binds another decision to this one, creating a new decision.
    ///
    /// Events accumulate in order, and a rejection terminates the chain.
    pub fn and_then<B, F: FnOnce(A) -> Decision<R, E, B>>(self, f: F) -> Decision<R, E, B> {
        match self {
            Decision::Accepted { events, result } => match f(result) {
                Decision::Accepted {
                    events: events2,
                    result,
                } => Decision::Accepted {
                    events: events.concat(events2),
                    result,
                },
                Decision::InDecisive(result) => Decision::Accepted { events, result },
                Decision::Rejected(reasons) => Decision::Rejected(reasons),
            },
            Decision::InDecisive(result) => f(result),
            Decision::Rejected(reasons) => Decision::Rejected(reasons),
        }
    }

    /// Alias of [`Decision::and_then`].
    pub fn flat_map<B, F: FnOnce(A) -> Decision<R, E, B>>(self, f: F) -> Decision<R, E, B> {
        self.and_then(f)
    }

    /// Sequences `next` after this decision, ignoring this output.
    pub fn then<B>(self, next: Decision<R, E, B>) -> Decision<R, E, B> {
        self.and_then(|_| next)
    }

    /// Replaces the output value.
    pub fn replace<B>(self, b: B) -> Decision<R, E, B> {
        self.map(|_| b)
    }

    /// Runs `f` for its events and rejections only, keeping this output.
    pub fn flat_tap<B, F: FnOnce(&A) -> Decision<R, E, B>>(self, f: F) -> Decision<R, E, A> {
        self.and_then(|a| f(&a).replace(a))
    }

    /// Whether this decision is rejected.
    pub fn is_rejected(&self) -> bool {
        matches!(self, Decision::Rejected(_))
    }

    /// Whether this decision is accepted.
    pub fn is_accepted(&self) -> bool {
        matches!(self, Decision::Accepted { .. })
    }

    /// Whether this decision is indecisive.
    pub fn is_indecisive(&self) -> bool {
        matches!(self, Decision::InDecisive(_))
    }

    /// Folds this decision: runs `fr` if it is rejected and `fa` otherwise.
    pub fn visit<B, FR, FA>(self, fr: FR, fa: FA) -> B
    where
        FR: FnOnce(NonEmpty<R>) -> B,
        FA: FnOnce(A) -> B,
    {
        match self {
            Decision::InDecisive(a) => fa(a),
            Decision::Accepted { result, .. } => fa(result),
            Decision::Rejected(reasons) => fr(reasons),
        }
    }

    /// Ignores events and converts this decision to a `Result`.
    pub fn to_result(self) -> Result<A, NonEmpty<R>> {
        self.visit(Err, Ok)
    }

    /// Ignores events and converts this decision to a `Result`, by reference.
    pub fn as_result(&self) -> Result<&A, &NonEmpty<R>> {
        match self {
            Decision::InDecisive(a) => Ok(a),
            Decision::Accepted { result, .. } => Ok(result),
            Decision::Rejected(reasons) => Err(reasons),
        }
    }

    /// Ignores events and errors and returns the program output, if any.
    pub fn to_option(self) -> Option<A> {
        self.visit(|_| None, Some)
    }

    /// The output value, if this decision is not rejected.
    pub fn result(&self) -> Option<&A> {
        self.as_result().ok()
    }

    /// The accepted events, if any.
    pub fn events(&self) -> Option<&NonEmpty<E>> {
        match self {
            Decision::Accepted { events, .. } => Some(events),
            _ => None,
        }
    }

    /// The rejection reasons, if this decision is rejected.
    pub fn rejections(&self) -> Option<&NonEmpty<R>> {
        self.as_result().err()
    }

    /// Ignores the output value.
    pub fn void(self) -> Decision<R, E, ()> {
        self.map(|_| ())
    }

    /// Validates the output with a function that may reject with several
    /// reasons.
    pub fn validate_with<B, F>(self, f: F) -> Decision<R, E, B>
    where
        F: FnOnce(A) -> Result<B, NonEmpty<R>>,
    {
        self.and_then(|a| Decision::validate(f(a)))
    }

    /// Validates the output with a function that may reject with a single
    /// reason.
    pub fn validate_one<B, F>(self, f: F) -> Decision<R, E, B>
    where
        F: FnOnce(A) -> Result<B, R>,
    {
        self.and_then(|a| Decision::from_result(f(a)))
    }

    /// Asserts the output with a function that may reject with several
    /// reasons, keeping the output unchanged.
    pub fn assert_with<B, F>(self, f: F) -> Decision<R, E, A>
    where
        F: FnOnce(&A) -> Result<B, NonEmpty<R>>,
    {
        self.flat_tap(|a| Decision::validate(f(a)))
    }

    /// Asserts the output with a function that may reject with a single
    /// reason, keeping the output unchanged.
    pub fn assert_one<B, F>(self, f: F) -> Decision<R, E, A>
    where
        F: FnOnce(&A) -> Result<B, R>,
    {
        self.flat_tap(|a| Decision::from_result(f(a)))
    }

    /// Recovers from a rejection.
    pub fn handle_error_with<F>(self, f: F) -> Decision<R, E, A>
    where
        F: FnOnce(NonEmpty<R>) -> Decision<R, E, A>,
    {
        match self {
            Decision::Rejected(reasons) => f(reasons),
            other => other,
        }
    }

    /// Stack-safe iteration: repeatedly applies `f` while it returns
    /// `Continue`, accumulating events along the way.
    ///
    /// This is the counterpart of Cats' `tailRecM`.
    pub fn tail_rec<S, F>(init: S, mut f: F) -> Decision<R, E, A>
    where
        F: FnMut(S) -> Decision<R, E, ControlFlow<A, S>>,
    {
        let mut acc: Vec<E> = Vec::new();
        let mut state = init;
        loop {
            match f(state) {
                Decision::Accepted { events, result } => match result {
                    ControlFlow::Continue(s) => {
                        acc.extend(events);
                        state = s;
                    }
                    ControlFlow::Break(b) => {
                        let mut events = events;
                        events.prepend(acc);
                        return Decision::Accepted { events, result: b };
                    }
                },
                Decision::InDecisive(result) => match result {
                    ControlFlow::Continue(s) => state = s,
                    ControlFlow::Break(b) => {
                        return match NonEmpty::from_vec(acc) {
                            None => Decision::InDecisive(b),
                            Some(events) => Decision::Accepted { events, result: b },
                        };
                    }
                },
                Decision::Rejected(reasons) => return Decision::Rejected(reasons),
            }
        }
    }

    /// Folds the output into `init` unless this decision is rejected.
    pub fn fold_left<B, F: FnOnce(B, &A) -> B>(&self, init: B, f: F) -> B {
        match self.as_result() {
            Ok(a) => f(init, a),
            Err(_) => init,
        }
    }

    /// Converts the events to another type.
    pub fn map_events<E2, F: FnMut(E) -> E2>(self, f: F) -> Decision<R, E2, A> {
        match self {
            Decision::InDecisive(a) => Decision::InDecisive(a),
            Decision::Accepted { events, result } => Decision::Accepted {
                events: events.map(f),
                result,
            },
            Decision::Rejected(reasons) => Decision::Rejected(reasons),
        }
    }

    /// Converts the rejection reasons to another type.
    pub fn map_rejections<R2, F: FnMut(R) -> R2>(self, f: F) -> Decision<R2, E, A> {
        match self {
            Decision::InDecisive(a) => Decision::InDecisive(a),
            Decision::Accepted { events, result } => Decision::Accepted { events, result },
            Decision::Rejected(reasons) => Decision::Rejected(reasons.map(f)),
        }
    }
}

impl<R, E> Decision<R, E, ()> {
    /// Constructs a program with a trivial output.
    pub fn unit() -> Self {
        Decision::InDecisive(())
    }

    /// Constructs a program that decides to accept a sequence of events.
    ///
    /// A single event or a [`NonEmpty`] of events can be passed.
    pub fn accept(events: impl Into<NonEmpty<E>>) -> Self {
        Decision::Accepted {
            events: events.into(),
            result: (),
        }
    }

    /// Constructs a program that accepts a sequence of events when
    /// `predicate` holds, or does nothing otherwise.
    pub fn accept_when(predicate: bool, events: impl Into<NonEmpty<E>>) -> Self {
        if predicate {
            Self::accept(events)
        } else {
            Self::unit()
        }
    }

    /// Constructs a program that accepts an optional event or does nothing.
    pub fn accept_some(event: Option<E>) -> Self {
        match event {
            Some(e) => Self::accept(e),
            None => Self::unit(),
        }
    }

    /// Constructs a program that accepts an optional event or rejects
    /// otherwise.
    pub fn accept_some_or(event: Option<E>, or_else: impl Into<NonEmpty<R>>) -> Self {
        match event {
            Some(e) => Self::accept(e),
            None => Self::reject(or_else),
        }
    }

    /// Constructs a program that accepts an event or rejects with a single
    /// reason.
    pub fn accept_ok(event: Result<E, R>) -> Self {
        match event {
            Ok(e) => Self::accept(e),
            Err(r) => Self::reject(r),
        }
    }

    /// Constructs a program that accepts an event or rejects with several
    /// reasons.
    pub fn accept_ok_nec(event: Result<E, NonEmpty<R>>) -> Self {
        match event {
            Ok(e) => Self::accept(e),
            Err(rs) => Decision::Rejected(rs),
        }
    }

    /// Constructs a program that rejects with a sequence of reasons when
    /// `predicate` holds, or does nothing otherwise.
    pub fn reject_when(predicate: bool, reasons: impl Into<NonEmpty<R>>) -> Self {
        if predicate {
            Self::reject(reasons)
        } else {
            Self::unit()
        }
    }

    /// Constructs a program that rejects with an optional single reason or
    /// does nothing otherwise.
    pub fn reject_some(reason: Option<R>) -> Self {
        match reason {
            Some(r) => Self::reject(r),
            None => Self::unit(),
        }
    }
}

impl<R, E, A> Decision<R, E, Option<A>> {
    /// Turns a decision of an optional output into an optional decision,
    /// keeping events and rejections (the `Traverse` instance for `Option`).
    pub fn transpose(self) -> Option<Decision<R, E, A>> {
        match self {
            Decision::InDecisive(None) | Decision::Accepted { result: None, .. } => None,
            Decision::InDecisive(Some(a)) => Some(Decision::InDecisive(a)),
            Decision::Accepted {
                events,
                result: Some(a),
            } => Some(Decision::Accepted { events, result: a }),
            Decision::Rejected(reasons) => Some(Decision::Rejected(reasons)),
        }
    }
}

impl<R, E, A, X> Decision<R, E, Result<A, X>> {
    /// Turns a decision of a fallible output into a fallible decision,
    /// keeping events and rejections (the `Traverse` instance for `Result`).
    pub fn transpose(self) -> Result<Decision<R, E, A>, X> {
        match self {
            Decision::InDecisive(Err(x)) | Decision::Accepted { result: Err(x), .. } => Err(x),
            Decision::InDecisive(Ok(a)) => Ok(Decision::InDecisive(a)),
            Decision::Accepted {
                events,
                result: Ok(a),
            } => Ok(Decision::Accepted { events, result: a }),
            Decision::Rejected(reasons) => Ok(Decision::Rejected(reasons)),
        }
    }
}

/// Sequences an iterator of decisions into a decision of a collection
/// (`traverse`/`sequence`): events accumulate in order and the first
/// rejection short-circuits.
///
/// ```
/// use edomata_core::{Decision, nonempty};
///
/// let ds = vec![Decision::accept_return(1, "a"), Decision::pure(2), Decision::accept_return(3, "b")];
/// let all: Decision<(), &str, Vec<i32>> = ds.into_iter().collect();
/// assert_eq!(all, Decision::Accepted { events: nonempty!["a", "b"], result: vec![1, 2, 3] });
/// ```
impl<R, E, A, C: Default + Extend<A>> FromIterator<Decision<R, E, A>> for Decision<R, E, C> {
    fn from_iter<I: IntoIterator<Item = Decision<R, E, A>>>(iter: I) -> Self {
        let mut events: Vec<E> = Vec::new();
        let mut out = C::default();
        for d in iter {
            match d {
                Decision::InDecisive(a) => out.extend([a]),
                Decision::Accepted {
                    events: evs,
                    result,
                } => {
                    events.extend(evs);
                    out.extend([result]);
                }
                Decision::Rejected(reasons) => return Decision::Rejected(reasons),
            }
        }
        match NonEmpty::from_vec(events) {
            None => Decision::InDecisive(out),
            Some(events) => Decision::Accepted {
                events,
                result: out,
            },
        }
    }
}

impl<R, E, A> From<Result<A, NonEmpty<R>>> for Decision<R, E, A> {
    fn from(result: Result<A, NonEmpty<R>>) -> Self {
        Decision::validate(result)
    }
}
